### ISSUE TITLE
Needs to stop watchdog in puppet implementation

### ISSUE THREAD

**Issue by windmemory** [wechaty/wechaty-puppet#106]
Is there a way to stop the `watchdog` in puppet implementation? I want to stop the `watchdog` in `wechaty-puppet-hostie` so I can deal with reconnection logic when the server is temporary unavailable.

**Comment by huan** [comment]
Of course.

You can make your dog sleep by calling: `dog.sleep()`.

See: https://github.com/huan/watchdog/blob/master/src/watchdog.ts#L238

**Comment by windmemory** [comment]
But `watchdog` is a `private` property of `wechaty-puppet`, it can not be accessed from puppet implementations. Shall we make it `protected`?

**Comment by huan** [comment]
Oh, yes it should be private.

I remembered that there's another way to stop the dog: feed it some poison foods...

Let me try to find it out...

**Comment by huan** [comment]
After I reviewed the code, I believe the `watchdog` should not be stopped, even you are doing a `reset` operation.

The reason why it should not be stopped is that when you are doing a `reset`, there's also have a risk of we get out of the control of our code, which should be monitored by our watchdog too.

So my suggestion would be: 

#### When you are doing a reset process, you can create a timer to feed the dog by emitting events that it requires.

Please let me know if you think this solution can be accepted, or we can continue discussing it.

BTW: I'm thinking about to refactor the watchdog system in the past months. So any workaround would be acceptable.

**Comment by windmemory** [comment]
The reason that I raise up this issue is because I was adding the reconnect logic in `wechaty-puppet-hostie`.

The `watchdog` is not initially active, for example, during the start phase of `wechaty-puppet-hostie`, if the grpc client failed to connect, there will be no any heartbeat events, but the watchdog won't complain, I think there is a mechanism that the watchdog won't work until the first food is feed.

But if the puppet is reset from some reason, and the grpc client can not be connected, during the reconnect phase, the reset will be triggered, and there will be double reconnect process, this is why I came up with this issue to discuss if we can stop the watchdog in puppet implementation.

**Comment by huan** [comment]
I understand that we need to deal with watchdog related states when enhancing the reconnect logic for our `wechaty-puppet-hostie`.

> The watchdog is not initially active, for example, during the start phase of wechaty-puppet-hostie, if the grpc client failed to connect, there will be no any heartbeat events, but the watchdog won't complain, I think there is a mechanism that the watchdog won't work until the first food is feed.

Yes, your analytics is absolutely right. My suggestion is that we can feed the watchdog inside the start call. The watchdog would be active after it has been feed for the first time.

> But if the puppet is reset from some reason, and the grpc client can not be connected, during the reconnect phase, the reset will be triggered, and there will be a double reconnecting process, this is why I came up with this issue to discuss if we can stop the watchdog in puppet implementation.

Yes, that's the problem that I was facing when I was writing the `wechaty-puppet-puppeteer`.

My solution is to save the `reset` status in another variable, and when we call `reset()`, we need to check whether there's another resetting task is running, and if so, the reset should deal this situation right.

I believe the above suggestion would be able to solve your problem, please feel free to let me know if I have missed anything.

**Comment by windmemory** [comment]
The thing that really bothers me is that I can not fully stop my puppet.

I would consider before start, the puppet is in `off` state, and watchdog is not active in `off` state. But if we do a reset, which means a stop then a start, we can not bring the puppet back to the `off` state.

**Comment by huan** [comment]
Yes, I understand.

As you suggested before, I believe the most convenient way for you now is to change the `watchdog` in `Puppet` from `private` to `protected`, am I right?

If so, please go ahead and feel free to submit a PR to change it as you want, I'm OK with it.

**Comment by windmemory** [comment]
Cool, will submit a PR later.

**Comment by windmemory** [comment]
I went through the whole issue again, and I think the best solutions should be that the `wechaty-puppet` has it's own `start` and `stop` implementation, so we can manage the watchdog state in the `start` and `stop` method. All puppet implementation could override the implementation or don't have to declare it at all. But this would introduce breaking changes to all puppets.

**Comment by huan** [comment]
I think your override solution is great.

And I can not understand why it will introduce breaking changes: they did not exists before so I think there will no breaking changes at all?

**Comment by windmemory** [comment]
> And I can not understand why it will introduce breaking changes: they did not exists before so I think there will no breaking changes at all?

Because currently the `start` and `stop` is an abstract method. So if we change it to a real method, will this break the current logic?

**Comment by huan** [comment]
Could you please explain that why changing a abstract method to real method will introduce a breaking change?

**Comment by windmemory** [comment]
I am not quite sure whether it will introduce breaking change or not, so would like to know your opinion.

**Comment by windmemory** [comment]
I went through the code in `wechaty-puppet`, came up with a solution below:

### Solution

The code below is in the `wechaty-puppet` `constructor()` function, I think we should put it into `start()` function, so we can also clear these resources in the `stop()` function.

```ts
    this.on('heartbeat', payload => this.watchdog.feed(payload))

    this.watchdog.on('reset', lastFood => {
      log.warn('Puppet', 'constructor() watchdog.on(reset) reason: %s', JSON.stringify(lastFood))
      this.emit('reset', lastFood)
    })

    /**
     * 2. Setup `reset` Event via a 1 second Throttle Queue:
     */
    this.resetThrottleQueue = new ThrottleQueue<string>(1000)
    // 2.2. handle all `reset` events via the resetThrottleQueue
    this.on('reset', payload => {
      log.silly('Puppet', 'constructor() this.on(reset) payload: "%s"', JSON.stringify(payload))
      this.resetThrottleQueue.next(payload.data)
    })
    // 2.3. call reset() and then ignore the following `reset` event for 1 second
    this.resetThrottleQueue.subscribe(reason => {
      log.silly('Puppet', 'constructor() resetThrottleQueue.subscribe() reason: "%s"', reason)
      this.reset(reason)
    })
```

### Issue

If we move these code from `constructor()` to `start()`, the current puppet implementation will need to add `super.start()` and `super.stop()` into their corresponding functions. If they don't have it, some resources won't be initialized. So this will be the breaking change.

---

I think the above solution is the ideal implementation for our `wechaty-puppet`, but we need to move to it gradually, so we can do this:
1. add an implementation for `start()` and `stop()` function without moving those initialize code from `constructor()` to `start()`. 
1. update all current puppets with `super.stop()` and `super.start()`. 
1. move logic from `constructor()` to `start()`.

@huan Let me know what do you think about it.

**Comment by huan** [comment]
Thanks for your solution proposal, I feel that it's the right direction we should follow.

In order to implement your code movements, I agree with you that we need to add `super.start()` and `super.stop()` in our Puppet Provider classes.

In addition, I think we can make the upgrade process to be very easy by just bumping the minor version of our `WechatyPuppet` class to the next developing version so that the current Puppet Providers classes should not be affected by the new changes at all.

Please feel free to go-ahead to implement your idea when you are comfortable with it.

**Comment by windmemory** [comment]
@huan I just submitted a PR related to the topics we discussed above, and I think we can close this issue after we merge that PR :P

**Comment by windmemory** [comment]
Let's close this issue since we have perfectly solved the problem with #108
